packages/mdc-textfield/bottom-line/foundation.js

```diff
@@ -0,0 +1,112 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import MDCFoundation from '@material/base/foundation';
+import MDCTextFieldBottomLineAdapter from './adapter';
+import {cssClasses, strings} from './constants';
+
+
+/**
+ * @extends {MDCFoundation<!MDCTextFieldBottomLineAdapter>}
+ * @final
+ */
+class MDCTextFieldBottomLineFoundation extends MDCFoundation {
+  /** @return enum {string} */
+  static get cssClasses() {
+    return cssClasses;
+  }
+
+  /** @return enum {string} */
+  static get strings() {
+    return strings;
+  }
+
+  /**
+   * {@see MDCTextFieldBottomLineAdapter} for typing information on parameters and return
+   * types.
+   * @return {!MDCTextFieldBottomLineAdapter}
+   */
+  static get defaultAdapter() {
+    return /** @type {!MDCTextFieldBottomLineAdapter} */ ({
+      addClass: () => {},
+      removeClass: () => {},
+      setAttr: () => {},
+      registerEventHandler: () => {},
+      deregisterEventHandler: () => {},
+      notifyOpacityTransitionEnd: () => {},
+    });
+  }
+
+  /**
+   * @param {!MDCTextFieldBottomLineAdapter=} adapter
+   */
+  constructor(adapter = /** @type {!MDCTextFieldBottomLineAdapter} */ ({})) {
+    super(Object.assign(MDCTextFieldBottomLineFoundation.defaultAdapter, adapter));
+
+    /** @private {function(!Event): undefined} */
+    this.transitionEndHandler_ = (evt) => this.transitionEnd(evt);
+  }
+
+  init() {
+    this.adapter_.registerEventHandler('transitionend', this.transitionEndHandler_);
+  }
+
+  destroy() {
+    this.adapter_.deregisterEventHandler('transitionend', this.transitionEndHandler_);
+  }
+
+  /**
+   * Activates the bottom line
+   */
+  activate() {
+    this.adapter_.addClass(cssClasses.BOTTOM_LINE_ACTIVE);
+  }
+
+  /**
+   * Animates out from the user's click location.
+   * @param {!Event} evt
+   */
+  animate(evt) {
+    const targetClientRect = evt.target.getBoundingClientRect();
+    const evtCoords = {x: evt.clientX, y: evt.clientY};
+    const normalizedX = evtCoords.x - targetClientRect.left;
+    const attributeString =
+      `transform-origin: ${normalizedX}px center`;
+
+    this.adapter_.setAttr('style', attributeString);
+  }
+
+  /**
+   * Deactives the bottom line
+   */
+  deactivate() {
+    this.adapter_.removeClass(cssClasses.BOTTOM_LINE_ACTIVE);
+  }
+
+  /**
+   * Fires when opacity transition ends, performing actions that must wait
+   * for the opacity animation to finish
+   * @param {!Event} evt
+   */
+  transitionEnd(evt) {
+    if (evt.propertyName === 'opacity') {
+      this.adapter_.notifyOpacityTransitionEnd();
```

Comment:
From my understanding, when the BottomLineFoundation receives a 'transitionend' event it tells the BottomLineAdapter to notify the TextFieldFoundation via a custom event, which then tells the BottomLineFoundation to deactivate(). 
Why can't we call deactivate() directly as part of the 'transitionend' handler in BottomLineFoundation?

Reply:
I cleaned up some of the method names to make this more clear. And here are some diagrams:

When a text field is "focused" (Aka the `mdc-text-field--focused` class is added), the parent foundation tells the bottom line to `activate`. That *also* causes the bottom line to emit an "animation end" event, but the parent foundation ignores it in this case:

![text field on focus](https://user-images.githubusercontent.com/4703432/32806399-3b4eef4c-c941-11e7-8262-7fb3b1da7587.png)

However, when a text field is "blurred" (Aka the `mdc-text-field--focused` class is remove), the bottom line foundation emits an "animation end" event. This is because removing `mdc-text-field--focused` changes the opacity of `mdc-text-field__bottom-line`. Once that event is emitted, the parent foundation listens to it and calls `deactivate` on the bottom line foundation.

![text field on blur](https://user-images.githubusercontent.com/4703432/32806449-6cf72ce4-c941-11e7-9a4f-5a7649f35696.png)